### PR TITLE
feat: [CDS-67838]: Separate Resource Interface for NgTemplateResource

### PR DESCRIPTION
**Pull request by utkarshchoubey-harness**
Separate Resource Interface and Impl for NgTemplateResource

Verified all the apis are shown in while running the services with develop branch

Verified basic template/Pipeline Operations :- 
* Template CRUD 
* Variables v2
* applyTemplates
* list-metadata
* Saving pipeline with templates
* move-config

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)

### CODE REVIEW

**Comment by utkarshchoubey-harness** [comment]
Work In Progress !
Made it ready for review to check if how many tests are failing

**Comment by utkarshchoubey-harness** [comment]
trigger messagecheck

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 32 of 75 files at r1, 5 of 12 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)* status: 77 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/TemplateServiceApplication.java` line 258 at r3](https://reviewable.io/reviews/harness/harness-core/47022#-NU1NZAY5p1u315ikoCI:-NU1NZAZ7MalXsRIYiR6:b-n0l1yq) ([raw file](https://github.com/harness/harness-core/blob/ec5eb102e736de0757c022ec867873d23a35dffd/840-template-service/src/main/java/io/harness/TemplateServiceApplication.java#L258)):*
> ```Java
> 
>   private void registerResources(Environment environment, Injector injector) {
>     Set<String> resources = new HashSet<>(List.of("io.harness.template.resources.NGTemplateResource"));
> ```

why do we need to do like this? we didn't do in other services



_Code quote:_
```Java
new HashSet<>(List.of("io.harness.template.resources.NGTemplateResource"))
```
___
*[`840-template-service/src/main/java/io/harness/template/resources/NGTemplateResourceImpl.java` line 478 at r3](https://reviewable.io/reviews/harness/harness-core/47022#-NU1NrkDE-ltoL0-AObo:-NU1NrkDE-ltoL0-AObp:b-btiver) ([raw file](https://github.com/harness/harness-core/blob/ec5eb102e736de0757c022ec867873d23a35dffd/840-template-service/src/main/java/io/harness/template/resources/NGTemplateResourceImpl.java#L478)):*
> ```Java
>   }
> 
>   public ResponseDTO<TemplateListRepoResponse> listRepos(@NotNull @AccountIdentifier String accountIdentifier,
> ```

ad @override annotation in all methods
___
*[`840-template-service/src/main/java/io/harness/template/services/TemplateRefreshServiceImpl.java` line 58 at r3](https://reviewable.io/reviews/harness/harness-core/47022#-NU1Nz6zC3qrFk3jLmmF:-NU1Nz6zC3qrFk3jLmmG:b-x5pn2o) ([raw file](https://github.com/harness/harness-core/blob/ec5eb102e736de0757c022ec867873d23a35dffd/840-template-service/src/main/java/io/harness/template/services/TemplateRefreshServiceImpl.java#L58)):*
> ```Java
>   private AccessControlClient accessControlClient;
> 
>   private static final String TEMPLATE = "TEMPLATE";
> ```

why do you need this?

**Review by prashantsharma04 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)* status: 77 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/TemplateServiceApplication.java` line 258 at r3](https://reviewable.io/reviews/harness/harness-core/47022#-NU1NZAY5p1u315ikoCI:-NU1ReAQ39Nat_gsz8gc:b32v57) ([raw file](https://github.com/harness/harness-core/blob/ec5eb102e736de0757c022ec867873d23a35dffd/840-template-service/src/main/java/io/harness/TemplateServiceApplication.java#L258)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

why do we need to do like this? we didn't do in other services
</blockquote></details>

+1
___
*[`840-template-service/contracts/src/main/java/io/harness/template/resources/beans/GetTemplateEntityRequest.java` line 8 at r4](https://reviewable.io/reviews/harness/harness-core/47022#-NU1RYxa87LPW1RxZ3vV:-NU1RYxa87LPW1RxZ3vW:bvu7nle) ([raw file](https://github.com/harness/harness-core/blob/da6043a7a7600aa15a7b5841e76bfb5d9a33f6d8/840-template-service/contracts/src/main/java/io/harness/template/resources/beans/GetTemplateEntityRequest.java#L8)):*
> ```Java
>  */
> 
> package io.harness.template.resources.beans;
> ```

Why is there an extra resource layer?

**Review by utkarshchoubey-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)* status: 77 file(s) to review. 2 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/template/resources/NGTemplateResourceImpl.java` line 478 at r3](https://reviewable.io/reviews/harness/harness-core/47022#-NU1NrkDE-ltoL0-AObo:-NU1aFSW2mJrAxNY06dC:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/ec5eb102e736de0757c022ec867873d23a35dffd/840-template-service/src/main/java/io/harness/template/resources/NGTemplateResourceImpl.java#L478)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

ad @override annotation in all methods
</blockquote></details>

Done.

**Review by utkarshchoubey-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)* status: 77 file(s) to review. 2 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/TemplateServiceApplication.java` line 258 at r3](https://reviewable.io/reviews/harness/harness-core/47022#-NU1NZAY5p1u315ikoCI:-NU5Oflo90Gw30rBc3Xn:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/ec5eb102e736de0757c022ec867873d23a35dffd/840-template-service/src/main/java/io/harness/TemplateServiceApplication.java#L258)):*
<details><summary><i>Previously, prashantsharma04 (Prashant Sharma) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.

**Review by utkarshchoubey-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)* status: 77 file(s) to review. 2 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/template/services/TemplateRefreshServiceImpl.java` line 58 at r3](https://reviewable.io/reviews/harness/harness-core/47022#-NU1Nz6zC3qrFk3jLmmF:-NU5V4ck345uBVs-Hmpl:b-r0hhtb) ([raw file](https://github.com/harness/harness-core/blob/ec5eb102e736de0757c022ec867873d23a35dffd/840-template-service/src/main/java/io/harness/template/services/TemplateRefreshServiceImpl.java#L58)):*
<details><summary><i>Previously, utkarshchoubey-harness wrote…</i></summary><blockquote>

It is used in Access Control check, Earlier it was used from beans, will add it to Constants and use it
</blockquote></details>

In the next pr I'll fix it for all resources

**Review by utkarshchoubey-harness — COMMENTED** [review]
> Reviewed 46 of 75 files at r1, 10 of 12 files at r2, 1 of 1 files at r3, 18 of 18 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)* status: 77 file(s) to review. 2 change requests. 3. unresolved discussions. 0 of 2 approvals obtained.

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 28 of 75 files at r1, 7 of 12 files at r2, 18 of 18 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)* status: 77 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/TemplateServiceApplication.java` line 258 at r3](https://reviewable.io/reviews/harness/harness-core/47022#-NU1NZAY5p1u315ikoCI:-NU5WKL1CBR1jzm93lMa:bw4f58x) ([raw file](https://github.com/harness/harness-core/blob/ec5eb102e736de0757c022ec867873d23a35dffd/840-template-service/src/main/java/io/harness/TemplateServiceApplication.java#L258)):*
<details><summary><i>Previously, utkarshchoubey-harness wrote…</i></summary><blockquote>

Done.
</blockquote></details>

@utkarshchoubey-harness what is the change here?

**Review by utkarshchoubey-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)* status: 77 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/TemplateServiceApplication.java` line 258 at r3](https://reviewable.io/reviews/harness/harness-core/47022#-NU1NZAY5p1u315ikoCI:-NU5_51X67rWm0PTT7VN:bggvl2c) ([raw file](https://github.com/harness/harness-core/blob/ec5eb102e736de0757c022ec867873d23a35dffd/840-template-service/src/main/java/io/harness/TemplateServiceApplication.java#L258)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

@utkarshchoubey-harness what is the change here?
</blockquote></details>

added the change

**Review by utkarshchoubey-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)* status: 77 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/TemplateServiceApplication.java` line 258 at r3](https://reviewable.io/reviews/harness/harness-core/47022#-NU1NZAY5p1u315ikoCI:-NU5_KoP0rkqN9FcY_yW:bdedyzx) ([raw file](https://github.com/harness/harness-core/blob/ec5eb102e736de0757c022ec867873d23a35dffd/840-template-service/src/main/java/io/harness/TemplateServiceApplication.java#L258)):*
<details><summary><i>Previously, utkarshchoubey-harness wrote…</i></summary><blockquote>

added the change
</blockquote></details>

removed the Resource.isAcceptable() check and in the getResources added the filter similar to what we have in other services

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)* status: 77 file(s) to review. 1 change requests. 1 of 2 approvals obtained.

**Review by prashantsharma04 — COMMENTED** [review]
> Reviewed 45 of 75 files at r1, 10 of 12 files at r2, 1 of 1 files at r3, 18 of 18 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)* status: 1 change requests. 1 of 2 approvals obtained.

File: 840-template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java [hunk 1]
```diff
@@ -169,21 +119,10 @@
...
-  private final TemplateMergeService templateMergeService;
-  private final VariablesServiceBlockingStub variablesServiceBlockingStub;
-  private final TemplateYamlConversionHelper templateYamlConversionHelper;
-  private final TemplateReferenceHelper templateReferenceHelper;
-  @Inject CustomDeploymentResourceClient customDeploymentResourceClient;
-  @Inject TemplateVariableCreatorFactory templateVariableCreatorFactory;
-
-  public static final String TEMPLATE_PARAM_MESSAGE = "Template Identifier for the entity";
+public interface NGTemplateResource {
+  String TEMPLATE = "TEMPLATE";
```
Comment (sribalij) [inline 840-template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java:123]: Wont need "TEMPLATE"?

Comment (utkarshchoubey-harness) [reply 840-template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java:123]: @sribalij have addressed this in the subsequent pr

**Review by prashantsharma04 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022#-:-NUPpLh-0CnYRQPmwGJX:bnfp4nl)* status: Some of the checks are too old. 1. unresolved discussions.

**Review by tchaurasiya — APPROVED** [review]
> Reviewed 45 of 75 files at r1, 10 of 12 files at r2, 1 of 1 files at r3, 18 of 18 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47022)* status: Some of the checks are too old. 1. unresolved discussions.

**Review by puthrayaharness — APPROVED** [review]

**Comment by utkarshchoubey-harness** [comment]
trigger review

**Comment by utkarshchoubey-harness** [comment]
trigger messagecheck

**Comment by utkarshchoubey-harness** [comment]
trigger review

**Comment by utkarshchoubey-harness** [comment]
trigger messagecheck

**Pull request merged**